components/collator/src/comparison.rs

```diff
@@ -1543,4 +1644,660 @@ impl CollatorBorrowed<'_> {
+                    // weights, which are not generated for alternate=non-ignorable.  There are
+                    // also exactly as many quaternary weights as tertiary weights, so level
+                    // length differences are handled already on tertiary level.  Any
+                    // above-common quaternary weight will compare greater regardless.
+                    quaternaries.append_byte(LEVEL_SEPARATOR_BYTE);
+                } else {
+                    let q = if q == NO_CE_QUATERNARY {
+                        LEVEL_SEPARATOR_BYTE
+                    } else {
+                        (0xfc + ((q >> 6) & 3)) as u8
+                    };
+                    handle_common!(
+                        quaternaries,
+                        q,
+                        common_quaternaries,
+                        QUAT_COMMON,
+                        QUAT_COMMON[WEIGHT_LOW]
+                    );
+                    quaternaries.append_byte(q);
+                }
+            }
+
+            if (non_primary.bits() >> 24) as u8 == LEVEL_SEPARATOR_BYTE {
+                break; // ce == NO_CE
+            }
+        }
+
+        macro_rules! write_level {
+            ($key:ident, $level:expr, $flag:ident) => {
+                if levels & $flag != 0 {
+                    if !need_to_write($level) {
+                        return;
+                    }
+                    sink.write(&[LEVEL_SEPARATOR_BYTE]);
+                    sink.write($key.as_ref());
+                }
+            };
+        }
+
+        write_level!(secondaries, Level::Secondary, SECONDARY_LEVEL_FLAG);
+
+        if levels & CASE_LEVEL_FLAG != 0 {
+            if !need_to_write(Level::Case) {
+                return;
+            }
+            sink.write(&[LEVEL_SEPARATOR_BYTE]);
+
+            // Write pairs of nibbles as bytes, except separator bytes as themselves.
+            let mut b = 0;
+            for c in cases.as_ref() {
+                assert_eq!(*c & 0xf, 0);
+                assert_ne!(*c, 0);
+                if b == 0 {
+                    b = *c;
+                } else {
+                    sink.write_byte(b | (*c >> 4));
+                    b = 0;
+                }
+            }
+            if b != 0 {
+                sink.write_byte(b);
+            }
+        }
+
+        write_level!(tertiaries, Level::Tertiary, TERTIARY_LEVEL_FLAG);
+        write_level!(quaternaries, Level::Quaternary, QUATERNARY_LEVEL_FLAG);
+    }
+}
+
+/// A [`std::io::Write`]-like trait for writing to a buffer-like object.
+///
+/// (This crate does not have access to [`std`].)
+pub trait Write {
+    /// Writes a buffer into the writer.
+    fn write(&mut self, buf: &[u8]);
+
+    /// Write a single byte into the writer.
+    fn write_byte(&mut self, b: u8) {
+        self.write(&[b]);
+    }
+
+    /// Write leading bytes up to a zero byte, but always write at least one byte.
```

Comment:
AFAICT, this is always called with an array of three bytes whose first byte is known to be non-zero, so do we need to check here for writing at least one byte?

Also, wouldn't it be better to have this as a helper function outside the trait so that trait implementors don't get to override this method with a broken implementation?

Reply:
Since there's one call site with at most 3 bytes and it already checks the first byte for zero, it seems to me that nested `if`s to check the other two bytes and writing each using a per-byte call would work.